### PR TITLE
R2D2 Feature extractor

### PR DESCRIPTION
**Pull request by lxxue**

### CODE REVIEW

File: hloc/extractors/r2d2.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+from ..utils.base_model import BaseModel
+
+r2d2_path = Path(__file__).parent / "../../third_party/r2d2"
+sys.path.append(str(r2d2_path))
+from extract import load_network, NonMaxSuppression, extract_multiscale
+
+RGB_mean = torch.cuda.FloatTensor([0.485, 0.456, 0.406])
+RGB_std  = torch.cuda.FloatTensor([0.229, 0.224, 0.225])
+norm_RGB = tvf.Normalize(mean=RGB_mean, std=RGB_std)
```
Comment (Skydes) [inline hloc/extractors/r2d2.py:14]: 1. This creates tensors on GPU even if this module is merely imported.
2. hloc should not break on GPU-only machines.

I suggest to:
- create `mean` and `std` in `_init` as `torch.tensor` (CPU) and register them with `register_buffer`. This way, they will be automatically moved to GPU with the model
- call `tvf.functional.normalize` in `_forward` (`mean` and `std` will already be on the correct device)

File: hloc/extractors/r2d2.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import sys
+from pathlib import Path
+import torch
+import torchvision.transforms as tvf
+
+from ..utils.base_model import BaseModel
+
+r2d2_path = Path(__file__).parent / "../../third_party/r2d2"
+sys.path.append(str(r2d2_path))
+from extract import load_network, NonMaxSuppression, extract_multiscale
```
Comment (Skydes) [inline hloc/extractors/r2d2.py:10]: This should require a `# noqa: E402` per `flake8`

File: hloc/extractors/r2d2.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        'top-k': 5000,
+
+        'scale-f': 2**0.25,
+        'min-size': 256,
+        'max-size': 1024,
+        'min-scale': 0,
+        'max-scale': 1,
+
+        'reliability-thr': 0.7,
+        'repetability-thr': 0.7,
```
Comment (Skydes) [inline hloc/extractors/r2d2.py:28]: 1. I usually use underscored instead of dashes for config entries
2. Could we replace `top-k` with `max_keypoints` to be consistent with `superpoint.py`
3. `scale-f` is not very explicit, is it `scale_factor`?
4. `thr` --> `threshold`

File: hloc/extractors/r2d2.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    def _init(self, conf):
+        model_fn = r2d2_path / "models" / conf['checkpoint_name']
+        self.net = load_network(model_fn)
+        self.detector = NonMaxSuppression(rel_thr=conf['reliability-thr'], 
+            rep_thr=conf['repetability-thr'])
+        
+    def _forward(self, data):
+        img = data['image']
+        img = norm_RGB(img[0])[None]
```
Comment (Skydes) [inline hloc/extractors/r2d2.py:40]: tvf normalization supports batching

File: hloc/extractors/r2d2.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        'reliability-thr': 0.7,
+        'repetability-thr': 0.7,
+    }
+    required_inputs = ['image']
+
+    def _init(self, conf):
+        model_fn = r2d2_path / "models" / conf['checkpoint_name']
+        self.net = load_network(model_fn)
+        self.detector = NonMaxSuppression(rel_thr=conf['reliability-thr'], 
+            rep_thr=conf['repetability-thr'])
```
Comment (Skydes) [inline hloc/extractors/r2d2.py:36]: This does not seem to respect PEP8

**Review by Skydes — CHANGES_REQUESTED** [review]
I added a few minor comments. Please make sure that the code respects the PEP8 style guidelines, for example by running `flake8`.

File: hloc/extractors/r2d2.py (hunk 1 above)
Comment (lxxue) [reply hloc/extractors/r2d2.py]: https://github.com/lxxue/Hierarchical-Localization/blob/01b71ee699832a96387e18dc4c70cb6212874772/hloc/extractors/r2d2.py#L28-L29

I pass the mean and std as list as required in this function(https://pytorch.org/docs/stable/_modules/torchvision/transforms/functional.html#normalize) and let the function itself to move the tensors to the suitable device. Is that ok?

File: hloc/extractors/r2d2.py (hunk 2 above)
Comment (lxxue) [reply hloc/extractors/r2d2.py:9]: All other feature extractors do the relative import like this. Should I add this `# noqa: E402` to all the models?

File: hloc/extractors/r2d2.py (hunk 3 above)
Comment (lxxue) [reply hloc/extractors/r2d2.py]: These variable names are set to be consistent with the original r2d2 repo. But I believe your suggestion of making it consistent with hloc is more reasonable.

**Comment by Matstah** [comment]
@lxxue How did you handle the descriptor dimension when using superglue for matching?

**Comment by lxxue** [comment]
@Matstah Sorry it doesn't support matching with superglue as they are trained on features of different dimensions. I think this is the same for other learned features like d2net.

**Comment by Skydes** [comment]
Thanks a lot for the changes!
